module/os/linux/zfs/zpl_xattr.c

```diff
@@ -729,19 +741,44 @@ __zpl_xattr_user_set(struct inode *ip, const char *name,
     const void *value, size_t size, int flags)
 {
 	char *xattr_name;
-	int error;
+	boolean_t compat = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_COMPAT);
+	boolean_t fallback = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_FALLBACK);
+	int error = 0;
 	/* xattr_resolve_name will do this for us if this is defined */
 #ifndef HAVE_XATTR_HANDLER_NAME
 	if (strcmp(name, "") == 0)
 		return (-EINVAL);
 #endif
+	if (ZFS_XA_NS_PREFIX_FORBIDDEN(name))
+		return (-EINVAL);
 	if (!(ITOZSB(ip)->z_flags & ZSB_XATTR))
 		return (-EOPNOTSUPP);
 
+	/*
+	 * Remove any namespaced version of the xattr so we only set the
+	 * version compatible with other platforms.
+	 *
+	 * The following flags must be handled correctly:
+	 *
+	 *   XATTR_CREATE: fail if xattr already exists
+	 *   XATTR_REPLACE: fail if xattr does not exist
+	 */
 	xattr_name = kmem_asprintf("%s%s", XATTR_USER_PREFIX, name);
-	error = zpl_xattr_set(ip, xattr_name, value, size, flags);
+	if (compat && fallback)
+		error = zpl_xattr_set(ip, xattr_name, NULL, 0, flags);
+	if (!compat)
+		error = zpl_xattr_set(ip, xattr_name, value, size, flags);
 	kmem_strfree(xattr_name);
 
+	if (!compat || error == -EEXIST)
+		return (error);
+	if (fallback && error == 0 && (flags & XATTR_REPLACE))
+		flags &= ~XATTR_REPLACE;
+	error = zpl_xattr_set(ip, name, value, size, flags);
+
+	dsl_dataset_t *ds = dmu_objset_ds(ITOZSB(ip)->z_os);
+	ds->ds_feature_activation[SPA_FEATURE_XATTR_COMPAT] = (void *)B_TRUE;
```

Comment:
Judging from the current uses of `ds_feature_activation` in `master`, I think it must only be accessed from syncing context. Otherwise, the following two problems occur:

**1. Data Race** between reader (syncing context) and writer (the entity setting `ds_feature_activation[fi] = ...`.
**2. Too-late activation** of the feature, i.e., `dsl_dataset_sync` observes `ds_feature_activation[fi] == B_TRUE` in a later txg than the `zpl_xattr_set` changes persist to disk, which require the feature to be activated on read.

Could someone familiar with the zfeature code please confirm or correct my assessment?

(cc'ing @pcd1193182 because he refactored it a few years ago, and @ahrens as the original author of zfeature)

Reply:
I think this is a good observation. I wonder if there is a way to activate the feature in the same txg? Or maybe activating it first and assuming success is the simplest option here...